e2e/benchmarks/services_test.go

```diff
@@ -0,0 +1,276 @@
+	srv := testutils.NewTestConsulServerHTTPS(b, "../../testutils")
+	defer srv.Stop()
+	path, cleanup := setupServiceInstances(b, srv, N)
+	defer cleanup() // comment out if you want to keep the config files
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		require.NoError(b, runSyncOnce(path))
+	}
+}
+
+// Benchmark N Services
+func Benchmark1Service(b *testing.B) {
+	benchmarkServices(b, 1)
+}
+func Benchmark100Service(b *testing.B) {
+	benchmarkServices(b, 1)
+}
+func Benchmark1000Service(b *testing.B) {
+	benchmarkServices(b, 1)
+}
+func benchmarkServices(b *testing.B, N int) {
+	srv := testutils.NewTestConsulServerHTTPS(b, "../../testutils")
+	defer srv.Stop()
+	path, cleanup := setupMultiServices(b, srv, N)
+	defer cleanup() // comment out if you want to keep the config files
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		require.NoError(b, runSyncOnce(path))
+	}
+}
+
+// ----------------------------------------------------------------------------
+
+// Run CTS in once mode from the cli level
+func runSyncOnce(configPath string) error {
+	cli := command.NewCLI(os.Stdout, os.Stderr)
+	args := []string{"cts", "-once", fmt.Sprintf("--config-file=%s", configPath)}
+	exitCode := cli.Run(args)
+	errstr := "Exit Code Erro: %s (%d)\n"
+	switch exitCode {
+	case command.ExitCodeOK:
+		return nil
+	case command.ExitCodeError:
+		return fmt.Errorf(errstr, "ExitCodeError", exitCode)
+	case command.ExitCodeInterrupt:
+		return fmt.Errorf(errstr, "ExitCodeInterrupt", exitCode)
+	case command.ExitCodeRequiredFlagsError:
+		return fmt.Errorf(errstr, "ExitCodeRequiredFlagsError", exitCode)
+	case command.ExitCodeParseFlagsError:
+		return fmt.Errorf(errstr, "ExitCodeParseFlagsError", exitCode)
+	case command.ExitCodeConfigError:
+		return fmt.Errorf(errstr, "ExitCodeConfigError", exitCode)
+	case command.ExitCodeDriverError:
+		return fmt.Errorf(errstr, "ExitCodeDriverError", exitCode)
+	default:
+		return fmt.Errorf(errstr, "UNKNOWN", exitCode)
+	}
+}
+
+// Sets up consul and config to test setup with N service instances
+func setupServiceInstances(t testing.TB, srv *testutil.TestServer, N int) (string, func() error) {
+	for _, s := range generateServices(N, true) {
+		testutils.RegisterConsulService(t, srv, s, testutil.HealthPassing)
+	}
+	return makeConfig(t, srv.HTTPAddr, false)
+}
+
+// add generated services as separte services
+func setupMultiServices(t testing.TB, srv *testutil.TestServer, N int) (string, func() error) {
+	for _, s := range generateServices(N, false) {
+		testutils.RegisterConsulService(t, srv, s, testutil.HealthPassing)
+	}
+	return makeConfig(t, srv.HTTPAddr, false)
+}
+
+// generate service instance entries
+func generateServices(n int, instances bool) []testutil.TestService {
+	baseport := 30000
+	services := make([]testutil.TestService, n)
+	for i := 0; i < n; i++ {
+		name, id := "service", fmt.Sprintf("svc_%d", i)
+		if instances {
+			name, id = fmt.Sprintf("svc_%d", i), "instance"
```

Comment:
I like how you're generating services vs. service instances here! It's very elegant